.gitlab-ci.yml

```diff
@@ -77,29 +77,23 @@ variables:
           --tag "$IMAGE_NAME:latest"
           --file "$DOCKERFILE" .
 
-check-typing:
-  stage:                           test
-  <<:                              *common-refs
-  <<:                              *kubernetes-env
-  script:
-    - yarn --immutable --network-concurrency 1
-    - yarn typing
-
 check-linting:
   stage:                           test
   <<:                              *common-refs
   <<:                              *kubernetes-env
   script:
-    - yarn --immutable --network-concurrency 1
-    - yarn lint
+    - apk add --no-cache python3 openssl make g++
```

Comment:
Is it really necessary to install these packages?

Reply:
Checked previous pipelines, I see why `python3` is needed. I can assume that `openssl` needs for ssl connections. But still can't understand about `make` and `g++`

Reply:
I guess they're for https://github.com/justmoon/node-bignum but I don't understand why this package is necessary, see my comment in https://github.com/paritytech/substrate-matrix-faucet/pull/150#discussion_r860737803.

https://github.com/justmoon/node-bignum/issues/114#issue-356377396 hints at the fact that we should probably use only the native `BigInt`, but I honestly haven't worked with Node.js numeric computing to have a strong opinion about this. Waiting to hear back from @mordamax.

Reply:
@alvicsam openssl is dep of node-bignum. I was adding them one by one, starting from openssl, then python3, it was still failing cuz wasn't able to find `make` etc.. however I tested locally with `gitlab-runner exec docker <command>` on my M1 chip, may be that could be a reason. It could be that it should work only with `openssl python3`, but on my machine this will still fail. Is there any issue with this to be added (so I better understand going forward)?

Reply:
@joao-paulo-parity re. numeric computing - honestly I was trying to avoid this dep as much as possible and tried tons of ways (that's why it was easier to make unit tests, so it's easier to play with it)
Here's the result of when I use only BigInt and direct multiplications

![image](https://user-images.githubusercontent.com/1177472/165758111-8a809cee-6e6e-4358-b612-15235972077a.png)

Reply:
```suggestion
    # TODO: remove after https://github.com/paritytech/ci_cd/issues/406
    - apk add --no-cache python3 openssl make g++
```

Okay, it's just better to keep minimum of additional packages in CI image. Also it seems that it's time to create a CI image for `substrate-matrix-faucet` as @joao-paulo-parity suggested.

Reply:
For the record I briefly tried https://github.com/davidmartinez10/bigfloat which is pure Node.js and it gave some different results

```
> yarn test
yarn run v1.22.18
$ jest
 FAIL  src/utils.spec.ts
  utils
    ✓ convertAmountToBn 0 => 0 (2 ms)
    ✓ convertAmountToBn 1 => 1000000000000 (1 ms)
    ✓ convertAmountToBn 100 => 100000000000000
    ✓ convertAmountToBn 1.45 => 1450000000000 (1 ms)
    ✓ convertAmountToBn 145.454141111 => 145454141111000
    ✕ convertAmountToBn 99999999.9 => 99999999000000000000 (2 ms)
    ✓ convertAmountToBn 999999999 => 999999999000000000000
    ✕ convertAmountToBn 999999999.9 => 999999999000000000000
    ✕ convertAmountToBn 999999999999999.9 => 999999999999999000000000000
    ✓ convertAmountToBn 999999999999999999999999999999999999999999999999999999 => 999999999999999999999999999999999999999999999999999999000000000000
    ✕ convertAmountToBn 999999999999999999999999999999999999999999999999999999.7 => 999999999999999999999999999999999999999999999999999999000000000000 (1 ms)

  ● utils › convertAmountToBn 99999999.9 => 99999999000000000000

    expect(received).toEqual(expected) // deep equality

    Expected: "99999999000000000000"
    Received: "99999999900000000000"

      24 |   ].forEach((t) =>
      25 |     test(`convertAmountToBn ${t.amount} => ${t.expected}`, () => {
    > 26 |       expect(convertAmountToBn(t.amount).toString()).toEqual(t.expected);
         |                                                      ^
      27 |     })
      28 |   );
      29 | });

      at Object.<anonymous> (src/utils.spec.ts:26:54)

  ● utils › convertAmountToBn 999999999.9 => 999999999000000000000

    expect(received).toEqual(expected) // deep equality

    Expected: "999999999000000000000"
    Received: "999999999900000000000"

      24 |   ].forEach((t) =>
      25 |     test(`convertAmountToBn ${t.amount} => ${t.expected}`, () => {
    > 26 |       expect(convertAmountToBn(t.amount).toString()).toEqual(t.expected);
         |                                                      ^
      27 |     })
      28 |   );
      29 | });

      at Object.<anonymous> (src/utils.spec.ts:26:54)

  ● utils › convertAmountToBn 999999999999999.9 => 999999999999999000000000000

    expect(received).toEqual(expected) // deep equality

    Expected: "999999999999999000000000000"
    Received: "999999999999999900000000000"

      24 |   ].forEach((t) =>
      25 |     test(`convertAmountToBn ${t.amount} => ${t.expected}`, () => {
    > 26 |       expect(convertAmountToBn(t.amount).toString()).toEqual(t.expected);
         |                                                      ^
      27 |     })
      28 |   );
      29 | });

      at Object.<anonymous> (src/utils.spec.ts:26:54)

  ● utils › convertAmountToBn 999999999999999999999999999999999999999999999999999999.7 => 999999999999999999999999999999999999999999999999999999000000000000

    expect(received).toEqual(expected) // deep equality

    Expected: "999999999999999999999999999999999999999999999999999999000000000000"
    Received: "999999999999999999999999999999999999999999999999999999700000000000"

      24 |   ].forEach((t) =>
      25 |     test(`convertAmountToBn ${t.amount} => ${t.expected}`, () => {
    > 26 |       expect(convertAmountToBn(t.amount).toString()).toEqual(t.expected);
         |                                                      ^
      27 |     })
      28 |   );
      29 | });

      at Object.<anonymous> (src/utils.spec.ts:26:54)

Test Suites: 1 failed, 1 total
Tests:       4 failed, 7 passed, 11 total
Snapshots:   0 total
Time:        2.236 s, estimated 3 s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```
import { BigFloat } from 'bigfloat.js'

export function convertAmountToBn(amount: string): BigFloat {
  const parsedAmount = new BigFloat(amount)

  return parsedAmount.mul(new BigFloat(10).pow(new BigFloat(decimals)))
}
```

What do you think @mordamax @chevdor?

Reply:
This looks nicer for sure :)  🥇
removes this ugly constant and too long comment

Reply:
It may be worth staying in line with the PolkadotJS API and use https://github.com/polkadot-js/common/tree/master/packages/x-bigint for the BigInt. Do you mind testing it ?
It is the one used by the API already so it will both limit the deps and stay in sync with the API.

Reply:
@chevdor I did test it, it's similar to regular BigInt, it seems more like a shim

Reply:
with BigFloat.js we were able 
- get rid of  ugly logic around max_float
- remove unnecessary openssl dependency and need to create an image
- add support of float decimals on big numbers

so It looks to be as preferable solution

Reply:
Ok, those are indeed sweet benefits for our use case but I don't see it very maintained with no change in 2 years.
If we go this route, we may want to depend on a fork we make so we can at least slide it some updates if required and try to get them upstream ?

# npm audit report

ansi-regex  4.0.0 - 4.1.0 || 5.0.0
Severity: high
 Inefficient Regular Expression Complexity in chalk/ansi-regex - https://github.com/advisories/GHSA-93q8-gq69-wqmw
 Inefficient Regular Expression Complexity in chalk/ansi-regex - https://github.com/advisories/GHSA-93q8-gq69-wqmw
fix available via `npm audit fix`
node_modules/ansi-regex
node_modules/string-width/node_modules/ansi-regex

glob-parent  <5.1.2
Severity: high
Regular expression denial of service in glob-parent - https://github.com/advisories/GHSA-ww39-953v-wcq6
fix available via `npm audit fix`
node_modules/glob-parent

lodash  <4.17.21
Severity: high
Command Injection in lodash - https://github.com/advisories/GHSA-35jh-r3h4-6jhm
fix available via `npm audit fix`
node_modules/lodash

minimist  <1.2.6
Severity: critical
Prototype Pollution in minimist - https://github.com/advisories/GHSA-xvch-5gv4-984h
fix available via `npm audit fix`
node_modules/minimist

4 vulnerabilities (3 high, 1 critical)

All of them can be fixed with `npm audit fix`.

Reply:
@chevdor good point, yea i think it makes sense, as current owner doesn't seem to be responsive there
do you think it makes sense do now? like creating a new issue for doing this

Reply:
I think we should. For a prod system with value I would say this is a must, but here for the faucet, we can make it easier on ourselves. It is always better to depend on the most maintained repo and if there is no such a repo, better use one of our own were we at least can patch things quickly as needed. If we patch the repo, we should obviously do a PR upsteam and **hope** the owner will take it. If (s)he does, great, if not, we know that the repo should not be counted on.

Reply:
I will approve in the current state but please let's track that down so we don't forget.